### PR TITLE
Polish analysis of a multiclosure test

### PR DESCRIPTION
**Pull request by comane**
Here we collect new functions and template that allow the analysis of multiclosure tests.
@comane @giovannidecrescenzo @mariaubiali @andreab1997

### CODE REVIEW

File: validphys2/src/validphys/kinematics.py
Comment (comane) [inline validphys2/src/validphys/kinematics.py:120]: @giovannidecrescenzo, @andreab1997  I think that this stuff should stay in `multiclosure_inconsistent.py` or `multiclosure_inconsistent_output.py`

Comment (comane) [reply validphys2/src/validphys/kinematics.py:120]: With this stuff, I mean all of the single datapoint ratio bias variance analysis

**Review by comane — COMMENTED** [review]
I think that the PR is ready for review.
The only things I would change as commented above is that the single data point ratio bias variance analysis does not reside in `kinematics.py` but rather in `multiclosure_inconsistent.py` / `multiclosure_inconsistent_output.py`

**Comment by andreab1997** [comment]
@comane thanks for this. If you want, ask myself for the review (but also add one between @scarlehoff and @RoyStegeman given that part of this PR is based on my work and it does not make sense for me to review my own functions )

File: validphys2/src/validphys/scripts/vp_multiclosure.py
Comment (comane) [inline validphys2/src/validphys/scripts/vp_multiclosure.py:15]:
```suggestion
import sys
import logging

from reportengine.compat import yaml

from validphys.app import App
from validphys import compareinconsistentclosuretemplates
```

These seem all unused

File: validphys2/src/validphys/compareinconsistentclosuretemplates/multiclosure_analysis.yaml
Comment (comane) [inline validphys2/src/validphys/compareinconsistentclosuretemplates/multiclosure_analysis.yaml:40]: Hi @andreab1997 , a weird thing here is that we can basically compare as many fits (see `lambdavalues` below) as we want in terms of pca_template, single_point_template, etc..

However, correct me if I am wrong, the part of this script that does `vp-comparefits` only compares two.

So if the above is right, I think that it would be nicer to have a `multi vp-comparefits` (meaning that more than 2 fits can be compared in principle) feature for this script.

File: validphys2/src/validphys/kinematics.py
Comment (comane) [inline validphys2/src/validphys/kinematics.py:24]: this module should probably not be touched at all

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [inline validphys2/src/validphys/closuretest/multiclosure.py:184]:
```suggestion
    closures_th, law_th, _, _ = internal_multiclosure_dataset_loader
    # The dimentions here are (fit, data point, replica)
    reps = np.asarray([th.error_members[:, :_internal_max_reps] for th in closures_th])
    # One could mask here some reps in order to avoid redundancy of information
    #TODO

    n_fits = np.shape(reps)[0]
    deltas = []
    # There are n_fits pdf_covariances
    # flag to see whether to eliminate dataset
    for i in range(n_fits):
        bias_diffs = np.mean(reps[i], axis = 1) - law_th.central_value
        # bias diffs in the for loop should have dim = (n_obs)
        sigmas = np.sqrt(np.var(reps[i], axis = 1))
        # var diffs has shape (n_obs,reps)
        bias_diffs = bias_diffs/sigmas
        
        deltas.append(bias_diffs.tolist())
        # biases.shape = (n_fits, n_obs_cut/uncut)
        # variances.shape = (n_fits, n_obs_cut/uncut, reps)
    return np.asarray(deltas)
```

These don't seem to be used anywhere

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [inline validphys2/src/validphys/closuretest/multiclosure.py:161]: would be good to have some numpydoc like docstrings

File: validphys2/src/validphys/closuretest/multiclosure_output.py
Comment (comane) [inline validphys2/src/validphys/closuretest/multiclosure_output.py:860]: some small description explaining what this does could be nice to have

**Review by comane — COMMENTED** [review]
Ciao @andreab1997 , I added some minor comments on some of the parts of the code that you and Giovanni wrote.

When you have the time, feel free to add a review to the `inconsistent_closuretest` folder

File: validphys2/examples/pca_bias_variance_report.yaml
Comment (andreab1997) [inline validphys2/examples/pca_bias_variance_report.yaml:8]: Why are you using the legacy versions? Is that necessary?

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py
Comment (andreab1997) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py:95]: I am wondering if it could be better to use `np.array` instead of lists

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py
Comment (andreab1997) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py:113]: Maybe this block can be put in another function since I guess it could be reused

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py
Comment (andreab1997) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py:57]: In general this function is too big for my taste, maybe it could be splitted in some more unit functions

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (andreab1997) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py:158]:
```suggestion

```

File: validphys2/src/validphys/closuretest/multiclosure_output.py
Comment (andreab1997) [inline validphys2/src/validphys/closuretest/multiclosure_output.py:81]: This block is used almost unchanged at least twice, maybe it would be sensible to have a unit function for that

File: validphys2/src/validphys/kinematics.py
Comment (andreab1997) [reply validphys2/src/validphys/kinematics.py:24]: I agree, could you remove the new lines?

**Review by andreab1997 — CHANGES_REQUESTED** [review]

**Comment by scarlehoff** [comment]
This is then ready to be merged? (@andreab1997 @comane )

**Comment by andreab1997** [comment]
> This is then ready to be merged? (@andreab1997 @comane )

I don't think so, surely I need to review this again but in any case I would wait for the paper to be published.

**Review by andreab1997 — COMMENTED** [review]
Ok I had a look and more or less nothing relevant changed since the last time I reviewed. @comane if you can answer the comments I left last time we can probably merge this soon.

File: validphys2/examples/pca_bias_variance_report.yaml
Comment (comane) [reply validphys2/examples/pca_bias_variance_report.yaml:8]: The fits that I use in that runcard were done using th 200. I think this is the reason, why I added the legacy version

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py
Comment (comane) [reply validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent.py]: Yes this makes sense to have perhaps. I added a function called `eigendecomposition` that returns the eigenvectors and eigenvalues orderered in descendent way

**Comment by comane** [comment]
Hi @andreab1997, I addressed most of the comments that you left, as well as those that I wrote myself.

There is one main issue with this PR at present, namely, the way in which the compare inconsistent closure test script works.

This script is problematic, because it's not possible to have a `dataset_inputs` which is different from `from_: fit`, note that this is a problem since we want to have out of sample datasets.

The problem, I think, can be solved by removing the vp-compare fits from `compareinconsistentclosuretemplates`.
We don't need to have another `vp-comparefits` anyways.

If you could take care of this that would be great.

**Comment by andreab1997** [comment]
> Hi @andreab1997, I addressed most of the comments that you left, as well as those that I wrote myself.
> 
> There is one main issue with this PR at present, namely, the way in which the compare inconsistent closure test script works.
> 
> This script is problematic, because it's not possible to have a `dataset_inputs` which is different from `from_: fit`, note that this is a problem since we want to have out of sample datasets.
> 
> The problem, I think, can be solved by removing the vp-compare fits from `compareinconsistentclosuretemplates`. We don't need to have another `vp-comparefits` anyways.
> 
> If you could take care of this that would be great.

Just to understand, this issue is only there if you use the CLI or even if you write your own runcard and template?

**Comment by comane** [comment]
> Just to understand, this issue is only there if you use the CLI or even if you write your own runcard and template?

I think it's there if I run `validphys multiclosure_analysis.yaml` (with dataset inputs that is not from the fit)

**Comment by andreab1997** [comment]
> > Just to understand, this issue is only there if you use the CLI or even if you write your own runcard and template?
> 
> I think it's there if I run `validphys multiclosure_analysis.yaml` (with dataset inputs that is not from the fit)

Ok I need to test this, maybe it is something fixable

**Comment by comane** [comment]
@scarlehoff, once the tests pass, this is ready to merge in my opinion

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:172]: Why don't you loop over `reps`, the `n_fits` variable doesn't seem necessary.

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:177]:
```suggestion
        bias_diffs /= sigmas
```

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:160]: Please add also the return variable to this docstr, smt like 

```
deltas: np.array
    array of shape (n_fits, nreps)
```
(or whatever the actual shape is)


If you add the input arguments as in the docstr of the function  below even better.

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:174]:
```suggestion
        # bias diffs in the for loop should have shape = (n_obs,)
```

Do you mean the shape? It is relevant for the comment below

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:181]: I guess variances are the sigmas? 

Also, I don't understand how the division work if they have different shapes?

(I guess some of the comments are from previous versions of the function? I like having many comments just want to make sure they are correct and updated!)

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:961]: Could you add some more info here. It's not clear why is it necessary.

e.g., could you use `xq2map_with_cuts` as the input? Instead of calling `xq2map_with_cuts`?

(maybe not)

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure.py:970]:
```suggestion
    # for the case of double-hadronic observables we have 2 (x,Q) for each experimental point
```

File: validphys2/src/validphys/closuretest/multiclosure_output.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure_output.py:118]: I don't understand why if you are creating the records in this same function you need to change the columns afterwards, can't you use the "right" names from the beginning? (instead of `dict` just use `{"error ratio" : <>}`, or does pandas complain if you do that?)

File: validphys2/src/validphys/closuretest/multiclosure_output.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/multiclosure_output.py:858]: Please add a docstring, I would not know what this function is doing :_

**Review by scarlehoff — COMMENTED** [review]
I haven't been able to test the code, but have had a look through and left a few comments.

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [reply validphys2/src/validphys/closuretest/multiclosure.py:177]: Again, a bit of code from @giovannidecrescenzo.

Here, I would rather have:
`delta = bias_diffs / sigma`

and then 

`delas.append(delta.tolist())`

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [reply validphys2/src/validphys/closuretest/multiclosure.py:181]: Yes, I agree, comments here make little sense. I now updated them.

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [reply validphys2/src/validphys/closuretest/multiclosure.py:961]: Thanks for the comment. 

xq2map_with_cuts is now taken in input and not loaded again

File: validphys2/src/validphys/closuretest/multiclosure_output.py
Comment (comane) [reply validphys2/src/validphys/closuretest/multiclosure_output.py:858]: I added a description. @giovannidecrescenzo just double check that it describes what the function is supposed to do please

File: validphys2/src/validphys/closuretest/runcards/old_rbv_definition.yaml
Comment (comane) [reply validphys2/src/validphys/closuretest/runcards/old_rbv_definition.yaml:56]: It's supposed to allow one to reproduce the results from the paper https://arxiv.org/pdf/2111.05787

The paper above used a "wrong" definition of Ratio Bias Variance. After the new closure test paper will be published I think that we can get rid of the old functions that were used for rbv, as well as of this runcard.

Because of this, I don't think that this belongs to examples

File: validphys2/examples/pca_bias_variance_report.yaml
Comment (comane) [reply validphys2/examples/pca_bias_variance_report.yaml:65]: smaller by reducing the number of fits that are included?
I think this would be reasonable yes

Comment (comane) [reply validphys2/examples/pca_bias_variance_report.yaml:65]: I left only 3

File: validphys2/src/validphys/closuretest/multiclosure.py
Comment (comane) [reply validphys2/src/validphys/closuretest/multiclosure.py:191]: yes you are right. I updated it now, thanks!

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py:501]:
```suggestion
    Returns
```

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py:375]:
```suggestion
    Returns
```

(or maybe the error is that you wanted to yield at the end?)

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (scarlehoff) [inline validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py:556]: Why do you need `df.columns` in this case if the names are not changing?

**Review by scarlehoff — COMMENTED** [review]
Thanks. I've now reviewed the part that I was missing (and ran a few reports).

I've tried to go through the entire thing but I haven't followed the project enough to be able to pick up any problems so take the review with that grain of salt, I haven't seen anything obvious though.

The only general question I have is, why do you need so many extra functionality under `inconsisent_closuretest`? I would have naively thought that many of the estimators could also be applied to the usual (multi)CT ?

As another request (maybe it is already in some of the other PRs), could you add a section in the docs for how to prepare and then analyze an inconsistent closure test? (hopefully together with the paper when sent to the arxiv)

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (comane) [reply validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py:166]: I am not sure that this works, as there is no `ds`

**Comment by comane** [comment]
> The only general question I have is, why do you need so many extra functionality under `inconsisent_closuretest`? I would have naively thought that many of the estimators could also be applied to the usual (multi)CT ?

I agree with you, the reason to do this is that the multiclosure.py module is already very large and full of unused functions.
I have now put all of the general multiclosure functions in multiclosure.py. 
In a future PR I will then polish the multiclosure module from the old and unused multiclosure functions.

File: validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py
Comment (comane) [reply validphys2/src/validphys/closuretest/inconsistent_closuretest/multiclosure_inconsistent_output.py]: Sure, it should have been taken care of now.

**Review by scarlehoff — APPROVED** [review]
Thanks for this, let's wait for the tests to finish then merge!

**Pull request merged**
